runtime/grid_sync.cu

```diff
@@ -220,7 +220,11 @@ __device__ void blockSerializeRelease(int64_t* semaphore) {
       index_utils::maskedOffset<X_BLOCK, Y_BLOCK, Z_BLOCK>(blockIdx, gridDim);
   bool last_block = block_idx_in_segment == segment_size - 1;
 
+  // Block until writes from all threads in this block are visible to all other
+  // blocks before releasing semaphore using thread 0.
+  __threadfence();
```

Comment:
Could you explain why this is necessary? An example would be appreciated.

Reply:
Added example in comment.

Reply:
Thanks. Makes sense.

Is __syncthreads still necessary?

Reply:
Yes they are both necessary because __threadfence does not synchronize the threads. It only causes that thread to wait until all its writes are visible globally. We need to sync so that all the threads finish their fences before thread 0 updates the semaphore.